tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+   * independent and not mutually exclusive. For instance, one could perform multilabel
+   * classification where a picture can contain both an elephant and a dog at the same time.
+   *
+   * <p>For brevity, let `x = logits`, `z = labels`. The logistic loss is
+   *
+   * <pre>
+   *     z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x))
+   *     = z * -log(1 / (1 + exp(-x))) + (1 - z) * -log(exp(-x) / (1 + exp(-x)))
+   *     = z * log(1 + exp(-x)) + (1 - z) * (-log(exp(-x)) + log(1 + exp(-x)))
+   *     = z * log(1 + exp(-x)) + (1 - z) * (x + log(1 + exp(-x))
+   *     = (1 - z) * x + log(1 + exp(-x))
+   *     = x - x * z + log(1 + exp(-x))
+   * </pre>
+   *
+   * <p>For x < 0, to avoid overflow in exp(-x), we reformulate the above
+   *
+   * <pre>
+   *      x - x * z + log(1 + exp(-x))
+   *      = log(exp(x)) - x * z + log(1 + exp(-x))
+   *      = - x * z + log(1 + exp(x))
+   * </pre>
+   *
+   * <p>Hence, to ensure stability and avoid overflow, the implementation uses this equivalent
+   * formulation
+   *
+   * <pre>
+   *     max(x, 0) - x * z + log(1 + exp(-abs(x)))
+   * </pre>
+   *
+   * <p>`logits` and `labels` must have the same type and shape.
+   *
+   * @param scope The TensorFlow scope
+   * @param labels the labels
+   * @param logits the logits of type float32 or float64
+   * @param <T> the type of labels and logits
+   * @return the component-wise logistic losses.
+   */
+  @Endpoint(name = "sigmoidCrossEntropyWithLogits")
+  public static <T extends TNumber> Operand<T> sigmoidCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<T> logits) {
+    if (labels.asOutput().shape().numDimensions() != logits.asOutput().shape().numDimensions())
+      throw new IllegalArgumentException(
+          String.format(
+              "logits and labels must have the same shape (%s vs %s)",
+              labels.asOutput().shape().toString(), logits.asOutput().shape()));
+    Operand<T> zeros =
+        Cast.create(scope, ZerosLike.create(scope, logits), logits.asOutput().dataType());
+    Operand<TBool> cond = GreaterEqual.create(scope, logits, zeros);
+
+    Operand<T> relu_logits = Select.create(scope, cond, logits, zeros);
+    Operand<T> neg_abs_logits = Select.create(scope, cond, Neg.create(scope, logits), logits);
+    return Add.create(
+        scope,
+        Sub.create(scope, relu_logits, Mul.create(scope, logits, labels)),
+        Log1p.create(scope, Exp.create(scope, neg_abs_logits)));
+  }
+
+  /**
+   * Flattens logits' outer dimensions and keep its last dimension.
+   *
+   * @param scope the TensorFlow scope
+   * @param logits the logits
+   * @param <T> the type of logits
+   * @return the flattened logits
+   */
+  private static <T extends TNumber> Operand<T> flattenOuterDims(Scope scope, Operand<T> logits) {
+    Operand<TInt64> one = Constant.scalarOf(scope, 1L);
+
+    org.tensorflow.ndarray.Shape shape = logits.asOutput().shape();
```

Comment:
you shouldn't need the full canonical name for `Shape` in this method as you've already imported it

Reply:
OK